src/hooks/useWatchHistory.ts

```diff
@@ -0,0 +1,124 @@
+ *
+ * @param player
+ * @param item
+ */
+export const useWatchHistory = (player: JWPlayer | undefined, item: PlaylistItem) => {
+  const queuedWatchProgress = useRef<QueuedProgress | null>(null);
+
+  // config
+  const { features } = useConfigStore((s) => s.config);
+  const continueWatchingList = features?.continueWatchingList;
+  const watchHistoryEnabled = !!continueWatchingList;
+
+  // watch History
+  const watchHistoryItem = useWatchHistoryStore((state) => (!!item && watchHistoryEnabled ? state.getItem(item) : undefined));
+
+  // update the queued watch progress on each time update event
+  const handleTimeUpdate = useEventCallback((event: jwplayer.TimeParam) => {
+    // live streams have a negative duration, we ignore these for now
+    if (event.duration < 0) return;
+
+    const progress = event.position / event.duration;
+
+    if (!isNaN(progress) && isFinite(progress)) {
+      queuedWatchProgress.current = {
+        item,
+        progress,
+      };
+    }
+  });
+
+  // maybe store the watch progress when we have a queued watch progress
+  const maybeSaveWatchProgress = useCallback(() => {
+    if (!watchHistoryEnabled || !queuedWatchProgress.current) return;
+
+    const { item, progress } = queuedWatchProgress.current;
+
+    // save the queued watch progress
+    saveItem(item, progress);
+
+    // clear the queue
+    queuedWatchProgress.current = null;
+  }, [watchHistoryEnabled]);
+
+  // listen for certain player events
+  useEffect(() => {
+    if (!player || !watchHistoryEnabled) return;
+
+    player.on('time', handleTimeUpdate);
+    player.on('pause', maybeSaveWatchProgress);
+    player.on('complete', maybeSaveWatchProgress);
+    player.on('remove', maybeSaveWatchProgress);
+
+    return () => {
+      player.off('time', handleTimeUpdate);
+      player.off('pause', maybeSaveWatchProgress);
+      player.off('complete', maybeSaveWatchProgress);
+      player.off('remove', maybeSaveWatchProgress);
+    };
+  }, [player, watchHistoryEnabled, maybeSaveWatchProgress, handleTimeUpdate]);
+
+  useEffect(() => {
+    return () => {
+      // store watch progress on unmount and when the media item changes
+      maybeSaveWatchProgress();
+    };
+  }, [item?.mediaid, maybeSaveWatchProgress]);
+
+  // add event listeners for unload and visibility change to ensure the latest watch progress is saved
+  useLayoutEffect(() => {
+    const visibilityListener = () => document.visibilityState === 'hidden' && maybeSaveWatchProgress();
+
+    window.addEventListener('beforeunload', maybeSaveWatchProgress);
+    window.addEventListener('visibilitychange', visibilityListener);
+
+    return () => {
+      window.removeEventListener('beforeunload', maybeSaveWatchProgress);
+      window.removeEventListener('visibilitychange', visibilityListener);
+    };
+  }, [maybeSaveWatchProgress]);
+
+  // calculate the `startTime` of the current item based on the watch progress
+  return useMemo(() => {
```

Comment:
I feel like this should probably be split into 2 separate hooks. In order to make sure that the listener hook always gets called, it's fine to call it inside of the useWatchHistory hook, but I think it would be nice to have everything from about approximately Lines 49 to 111 encapsulated in a useWatchHistoryListener hook.

Reply:
Done!